### PR TITLE
TinyMCE drag and drop bug fixes

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue ##14906 .

### Summary of Changes
Apart from fixing the bug in #14906 this PR also fixes another bug with the toolbar tooltip rendered in wrong position after the first image upload.
Also moves the plugin javascript code out of the tinyMCE/plugins directory (will make updates easier)

### Testing Instructions
Edit the configuration and provide a custom path for the drag and drop upload, this setting is per tab option so make sure that you are using the right one!!!
Try to upload an image,
Check where the image is stored and also try to see if the tooltips in bold, italics etc is in the right position

Apply the patch and repeat

### Expected result
Use the provided path

### Actual result
Is not using the provided path

### Documentation Changes Required
None, bug fix

### CODE REVIEW

**Comment by AlexRed** [comment]
ok, the patch fix the bug with the editor toolbar tooltip rendered in wrong position after the first image upload.
But for me the "Images Directory" parameter tooltip need to explain you can use the "Images Directory" parameter in the TinyMCE plugin to insert a folder name (not a path). This will be automatically generated by the system on the first dragging of an image. The folder will be created as a subfolder of the one set in the media manager settings as “Path to Files Folder”. Since they are image files, I think it would be better to set that it is created as a subfolder of the one set in the media manager settings as “Path to Images Folder”.

**Comment by dgt41** [comment]
The last commit transforms the inout to a drop down of possible directories, as @brianteeman suggested in the issue.
Personally I always prefer clicking instead of typing (although as a dev it should be the other way around :) )

screen shot 2017-04-02 at 19 57 12

For the tooltip text please make another PR

File: plugins/editors/tinymce/field/uploaddirs.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	 * @return  array  The skins option objects.
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function getOptions()
+	{
+		$options = array();
+		$root    = JComponentHelper::getParams('com_media')->get('image_path');
+
+
```
Comment (yvesh) [inline plugins/editors/tinymce/field/uploaddirs.php]: empty line :-)

File: plugins/editors/tinymce/field/uploaddirs.php [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+/**
+ * Generates the list of directories  available for drag and drop upload.
+ *
+ * @package     Joomla.Plugin
+ * @subpackage  Editors.tinymce
+ * @since       __DEPLOY_VERSION__
+ */
+class JFormFieldUploaddirs extends JFormFieldList
+{
+	protected $type = 'uploaddirs';
```
Comment (yvesh) [inline plugins/editors/tinymce/field/uploaddirs.php]: do comment missing

File: plugins/editors/tinymce/tinymce.php [hunk 3]
```diff
@@ -1877,15 +1883,15 @@ private function onDisplayLegacy($name, $content, $width, $height, $col, $row, $
...
-		var mediaUploadPath = '" . $tempPath . "';
-		var uploadUri       = '" . $uploadUrl . "';
-			"
+
+			$scriptOptions['setCustomDir']    = $isSubDir;
+			$scriptOptions['mediaUploadPath'] = $tempPath;
+			$scriptOptions['uploadUri']       = $uploadUrl;
+
+			$externalPlugins = array(
+				array('jdragdrop' => ($app->isClient('site') ? JUri::root(false)  : str_replace('/administrator', JUri::root(false)))
```
Comment (yvesh) [inline plugins/editors/tinymce/tinymce.php]: JUri::root() should never contain administrator? :-) 

Juri::root() in example.com/administrator returns example.com

File: plugins/editors/tinymce/tinymce.php [hunk 4]
```diff
@@ -573,7 +573,9 @@ public function onDisplay($name, $content, $width, $height, $col, $row, $buttons
 
 		if ($dragdrop && $user->authorise('core.create', 'com_media'))
 		{
-			$plugins[]     = 'jdragdrop';
+			$externalPlugins['jdragdrop'] = ($app->isClient('site') ? JUri::root(false)  :
+					str_replace('/administrator', '', JUri::root(false)))
```
Comment (yvesh) [inline plugins/editors/tinymce/tinymce.php]: JUri::root() should never contain administrator? :-)

Juri::root() in example.com/administrator returns example.com

**Comment by AlexRed** [comment]
I have tested this item :red_circle: unsuccessfully on 330dcc06a4091e2e4c93ee620b6d9bcbf005a28c<br /><br />no work for me.
I select "headers" in the dropdown and "images/draganddrop" in Path to Files Folder in the Media: Options.
The image go in the new folder   images/draganddrop/headers/<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/15057">issues.joomla.org/tracker/joomla-cms/15057</a>.</sub>

**Comment by AlexRed** [comment]
Sorry but I can't make a PR for the en language file..  my english is very bab  :(<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/15057">issues.joomla.org/tracker/joomla-cms/15057</a>.</sub>

**Comment by zero-24** [comment]
Please add a doc block for the new method to fix drone ;)

**Comment by dgt41** [comment]
@AlexRed thanks, that should be ok now

**Comment by AlexRed** [comment]
no more dropdown ?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/15057">issues.joomla.org/tracker/joomla-cms/15057</a>.</sub>

**Comment by dgt41** [comment]
It's still there for me:
screen shot 2017-04-02 at 20 56 33

**Comment by AlexRed** [comment]
sorry, dropdown ok but the image continue to go in a subfolder of Path to Files Folder in the Media: Options.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/15057">issues.joomla.org/tracker/joomla-cms/15057</a>.</sub>

**Comment by dgt41** [comment]
@AlexRed are you on M$ win-dos machine?

**Comment by AlexRed** [comment]
No, linux

**Comment by dgt41** [comment]
@AlexRed so can you post your settings, in tinymce and com_media, I can't really understand why your setup is misbehaving

**Comment by AlexRed** [comment]
yes, I can. But the Joomla users when select the folder in the dropdown think the images go there...  but no. The images go in a new subfolder with the same name if the user use the Path to Files Folder in the Media

**Comment by dgt41** [comment]
> But the Joomla users when select the folder in the dropdown think the images go there...

This folder that they will select is **always** a subfolder of their images directory, maybe a better tooltip there could be more helpful

**Comment by AlexRed** [comment]
yes, also for me the tooltip text need to explain better

**Comment by brianteeman** [comment]
use the esisting PLG_FIELDS_MEDIA_PARAMS_DIRECTORY_DESC

**Comment by dgt41** [comment]
@brianteeman done

**Comment by dgt41** [comment]
@brianteeman was that what you wanted to select in the issue tracker?

**Comment by brianteeman** [comment]
oops no

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 9e6415a74ca79ceb648eb716891387c2799edd20<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/15057">issues.joomla.org/tracker/joomla-cms/15057</a>.</sub>

**Comment by AlexRed** [comment]
I have tested this item :white_check_mark: successfully on 9e6415a74ca79ceb648eb716891387c2799edd20<br /><br />Patch ok for me<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/15057">issues.joomla.org/tracker/joomla-cms/15057</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
RTC after two successful testes.

**Pull request merged**